File: java/src/main/java/org/libpetri/core/Subnet.java
package org.libpetri.core;

/**
 * Sealed sum-type abstraction over Petri nets, distinguishing closed
 * (runnable) nets from open (composable) subnet fragments per
 * {@code spec/11-modular-composition.md}, requirement <b>MOD-002</b>.
 *
 * <p>This abstraction is the bridge between the existing flat {@link PetriNet}
 * world and the modular composition extension. Code that wishes to accept
 * "any net" types its parameter as {@code Subnet}; code that needs a runnable
 * net continues to take {@link PetriNet} (or {@link Closed}); code that needs
 * an open fragment takes {@link Open} (or {@link SubnetDef}{@code <?>}).
 * Misuse is rejected at compile time wherever the language permits.
 *
 * <h2>Variants</h2>
 * <ul>
 *   <li>{@link Closed} — wraps a {@link PetriNet}, runnable by an executor per
 *       [CORE-040].</li>
 *   <li>{@link Open} — implemented by {@link SubnetDef}{@code <P>} for any
 *       parameter type {@code P}; an open fragment with a declared
 *       {@link Interface} but no executor binding.</li>
 * </ul>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * Subnet asAny = new Subnet.Closed(myNet);
 * Subnet asOpen = SubnetDef.builder("Buffer").build();
 *
 * switch (asAny) {
 *     case Subnet.Closed(PetriNet net) -> runWithExecutor(net);
 *     case Subnet.Open open            -> compose(open);
 * }
 * }</pre>
 *
 * @see PetriNet
 * @see SubnetDef
 * @see Interface
 */
public sealed interface Subnet permits Subnet.Closed, Subnet.Open {

    /**
     * Closed (runnable) net variant. Wraps an existing {@link PetriNet}.
     *
     * @param net the underlying flat Petri net
     */
    record Closed(PetriNet net) implements Subnet {
        public Closed {
            if (net == null) {
                throw new IllegalArgumentException("Closed.net cannot be null");
            }
        }
    }

    /**
     * Open (composable) net variant. Implemented by {@link SubnetDef} —
     * the sealing of {@code Open} to a single permitted subtype is intentional:
     * subnet definitions are the sole open-net carrier in this design.
     */
    sealed interface Open extends Subnet permits SubnetDef { }
}

File: typescript/src/core/subnet.ts
import type { PetriNet } from './petri-net.js';
import type { SubnetDef } from './subnet-def.js';

/**
 * Discriminated sum-type abstraction over Petri nets, distinguishing **closed**
 * (runnable) nets from **open** (composable) subnet fragments per
 * `spec/11-modular-composition.md` requirement **MOD-002**.
 *
 * This abstraction is the bridge between the existing flat {@link PetriNet}
 * world and the modular composition extension. Code that wishes to accept
 * "any net" types its parameter as `Subnet`; code that needs a runnable net
 * continues to take {@link PetriNet} (or `Subnet` narrowed to `'closed'`);
 * code that needs an open fragment takes {@link SubnetDef} (or `Subnet`
 * narrowed to `'open'`). Misuse is rejected at compile time wherever
 * TypeScript's structural typing permits.
 *
 * Use the literal `kind` field for exhaustive pattern matching:
 *
 * ```ts
 * function classify(s: Subnet): string {
 *   switch (s.kind) {
 *     case 'closed': return `closed:${s.net.name}`;
 *     case 'open':   return `open:${s.def.name}`;
 *   }
 * }
 * ```
 */
export type Subnet =
  | { readonly kind: 'closed'; readonly net: PetriNet }
  | { readonly kind: 'open'; readonly def: SubnetDef<unknown> };

/** Wraps a {@link PetriNet} as a closed (runnable) {@link Subnet}. */
export function closedSubnet(net: PetriNet): Subnet {
  return { kind: 'closed', net };
}

/** Wraps a {@link SubnetDef} as an open (composable) {@link Subnet}. */
export function openSubnet(def: SubnetDef<unknown>): Subnet {
  return { kind: 'open', def };
}
